Streamlining and minor cleanup of code in the PrettyNaming module

jack-pappas opened the pull request:
I've made some small changes to the PrettyNaming module to streamline it for performance. Specifically:
* Compilation and decompilation of operators is now memoized for custom (non-built-in) operators to improve performance.
* The ``IsInfixOperator`` function was creating (potentially) multiple lists each time it was called. I've changed the lists to arrays since they're never modified, and I've lifted these arrays outside of the function body so they're only created once and re-used.
* Some repetitive ``if``-``then``-``else`` expressions have been changed to use ``match`` instead so they can be compiled into a ``switch`` opcode when appropriate.
* A few places in the code were unnecessarily using the ``.Substring(...)`` operation; they have been changed to use non-allocating comparisons instead (e.g., ``.StartsWith(...)``).
* The ``IsValidPrefixOperatorUse``, ``IsValidPrefixOperatorDefinitionName``, and ``IsPrefixOperator`` functions would, in certain specific cases, allocate a closure. It doesn't appear that this is actually necessary though -- a small modification to the code eliminates the variable capture that causes the allocation, allowing the lambda to be lifted out of the functions.

I've done some rudimentary benchmarking, and this change appears to have no significant impact (better or worse) when I compile these changes into a new ``fsc-proto`` and use that to rebuild the ``FSharp.Compiler.dll``. However, these changes do appear to improve compile times by a small but consistent amount when I use the ``fsc-proto`` to compile the various flavors of ``FSharp.Core``. Based on the data produced by ``--times``, these improvements appear to come from reducing the number of Gen0 and Gen1 collections during the ``[Parse inputs]`` and ``[Typecheck]`` compilation phases.

These modifications should not cause any externally-visible differences in behavior (i.e., from the perspective of callers in other parts of the compiler).

msftclas commented:
Hi __@jack-pappas__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

jack-pappas commented:
Do I need to sign the CLA again? I thought I'd already filed this one.

forki commented:
@jack-pappas there was a change of the CLA process a couple of months ago. the new one is much faster/easier to sign and doesn't need employers approval. did you sign this one?

dsyme commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +109,158 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+    let private decompileCustomOpName =
+        // Memoize this operation. Custom operators are typically used more than once
+        // so this avoids repeating decompilation.
+        let decompiledOperators = Dictionary<_,_> (System.StringComparer.Ordinal)
```
Please put this as a private module-level definition and mark it with // +++ GLOBAL STATE

dsyme replied:
Any reason not to use ConcurentDictionary?  I'd prefer that unless there's a proven case against it.

dsyme commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +109,158 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+    /// For example, "op_DereferencePercent" becomes "!%"
+    let DecompileOpName =
+        /// Maps the mangled operator names of built-in F# operators back to the operators.
+        let standardOps =
```
Likewise this one.

dsyme commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -177,59 +269,89 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
-        | _ -> s.[0] = '!' || (s.[0] = '~' && String.forall (fun c -> c = s.[0]) s)
+        | _ ->
+            s.[0] = '!'
+            || (String.forall (fun c -> c = '~') s)
```
I think the first ``s.[0] = '~'``  is voiding a call to String.forall, just leave it as it was for this one

dsyme commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -177,59 +269,89 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
-        DecompileOpName s = "?<-"
-
-    let IsInfixOperator s (* where s is assumed to be a compiled name *) =
+        DecompileOpName s = qmarkSet
```
Please put parentheses around this

dsyme commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -177,59 +269,89 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+
+    let IsInfixOperator =
+        /// EQUALS, INFIX_COMPARE_OP, LESS, GREATER
+        let relational = [| "=";"!=";"<";">";"$"|]
```
The compiler generally doesn't use the ``let f = let ... in fun x -> ...`` style.  There's no real reason why not except we generally hoist the state and/or amortized values to private module bindings.  I think it's best to hoist them explicitly here too even if they are not mutable

dsyme commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +109,158 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+            match standardOpNames.TryGetValue op with
+            | true, x -> x
+            | false, _ ->
+                if not <| IsOpName op then op
```
Please don't use ``<|`` in the compiler code base, at least not with ``not``.  It is not particularly readable.  There are occasional exceptions when providing function arguments to combinators

dsyme commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -177,59 +269,89 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+        /// INFIX_AT_HAT_OP
+        let concat = [| "@";"^" |]
+        /// PLUS_MINUS_OP, MINUS
+        let plus_minus = [| "+"; "-" |]
```
Please use `plusMinus`` - generally we don't use underscores in the compiler codebase

dsyme commented:
I've taken a close look and it all seems like reasonable cleanup.  But I'd appreciate someone else checking the code closely too.

dsyme commented:
I've left some comments too.

msftclas commented:
__@jack-pappas__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -293,7 +417,7 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
         res
 
-    type NameArityPair = NameArityPair of string*int
+    type NameArityPair = NameArityPair of name:string * arity:int
```
Named DU fields is a 3.1 feature, but our internal build uses a pre-3.1 compiler to bootstrap.  So for now we can't do this.

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -408,7 +532,7 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
 
     type ActivePatternInfo = 
-        | APInfo of bool * string list 
+        | APInfo of isTotal:bool * activeTags:string list
```
Same comment on named DU fields

jack-pappas commented:
@dsyme I've made some of your requested changes and updated this PR (I'll squash them before the PR is merged).

I wasn't able to make ``CompileOp`` and ``DecompileOp`` private to the module; when I did that and tried to compile, I got a number of errors:
```
"C:\visualfsharp\src\fsharp-proto-build.proj" (default target) (1) ->
"C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj" (Build target) (3) ->
(CoreCompile target) ->
  C:\visualfsharp\src\fsharp\ast.fs(1713,23): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\ast.fs(1714,19): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\ast.fs(1715,46): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\ast.fs(1726,37): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\pars.fsy(3684,74): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\pars.fsy(3689,93): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\pars.fsy(3692,95): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\pars.fsy(3696,93): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\pars.fsy(4611,36): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\pars.fsy(4614,118): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\pars.fsy(4617,36): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(837,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(838,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(839,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(840,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(842,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(844,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(845,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(846,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(847,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(848,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(849,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(850,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(851,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TcGlobals.fs(852,110): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TastOps.fs(3024,26): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\NameResolution.fs(200,38): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\NameResolution.fs(202,36): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\ConstraintSolver.fs(1172,100): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\ConstraintSolver.fs(1174,97): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\ConstraintSolver.fs(1183,36): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\ConstraintSolver.fs(1228,157): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\ConstraintSolver.fs(1230,154): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(971,20): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(1289,151): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(1290,137): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(1291,124): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(1393,52): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(5741,55): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(7092,27): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(7104,83): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(7126,83): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\TypeChecker.fs(7537,69): error FS0039: The value or constructor 'CompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\CompileOps.fs(720,47): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\CompileOps.fs(722,49): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\CompileOps.fs(724,24): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\CompileOps.fs(733,46): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]
  C:\visualfsharp\src\fsharp\CompileOps.fs(736,42): error FS0039: The value or constructor 'DecompileOpName' is not defined [C:\visualfsharp\src\fsharp\FSharp.Compiler-proto\FSharp.Compiler-proto.fsproj]

    0 Warning(s)
    48 Error(s)
```

I tried modifying the code to use ``ConcurrentDictionary<_,_>`` but ran into a bug in the compiler's type inferencer; I'll open an issue for this, but basically there's no way to call the ``ConcurrentDictionary<'TKey, 'T>.GetOrAdd('TKey, System.Func<'TKey, 'T>)`` overload. So, we can either stick with the dictionary + lock (for now, or permanently) or I can implement this in another way using ConcurrentDictionary which will have the downside of being slightly less efficient (when called concurrently for the same key); either way is fine with me, just let me know what you prefer.

jack-pappas commented:
@latkin I've removed the named fields and updated the code.

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -318,38 +458,35 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+            if s.StartsWith("get_", System.StringComparison.Ordinal) ||
+               s.StartsWith("set_", System.StringComparison.Ordinal)
+            then Some (s.Substring(4, s.Length - 4))
+            else None
```
Can we simplify this further, so there is only a single check for "get_" and "set_" prefixes?

```fsharp
    /// Try to chop "get_" or "set_" from a string
    let TryChopPropertyName (s: string) =
        // extract the logical name from any mangled name produced by MakeMemberDataAndMangledNameForMemberVal
        if s.Length <= 4 then None else
        let s = chopStringTo s '.'
        if s.StartsWith("get_", System.StringComparison.Ordinal) ||
           s.StartsWith("set_", System.StringComparison.Ordinal)
            then Some (s.Substring(4, s.Length - 4))
            else None
```

jack-pappas replied:
I was thinking about making this change when doing the cleanup. It's a good change, I'll make it now and update the PR.

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +109,162 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
-        let rec loop i = (i < n.Length && (opCharDict.ContainsKey(n.[i]) || loop (i+1)))
+    let IsOpName (name:string) =
+        let nameLen = name.Length
+        let rec loop i = (i < nameLen && (opCharSet.Contains(name.[i]) || loop (i+1)))
```
I think this would be much clearer as

```fsharp
let IsOpName (name:string) =
    name |> String.exists opCharDict.Contains
```

jack-pappas replied:
I agree, that would be much clearer. However -- would that end up allocating a delegate and/or FSharpFunc<_,_> each time ``IsOpName`` is called? Or will the compiler know to lift the creation of the delegate / FSharpFunc outside of ``IsOpName``?

latkin replied:
It would allocate a new fsharpfunc on each invocation

dsyme replied:
I can't recall if this is perf critical, but since it might well be I think it's better to leave it as it was

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +109,162 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+            let t2 = Dictionary<_,_> (opCharTranslateTable.Length)
+            for x, y in opCharTranslateTable do
+                t2.Add (x, y)
+            t2
```
Not sure if the `dict` function is defined/available here, but if it is, this would simplify to just `let t2 = dict opCharTranslateTable`

jack-pappas replied:
``dict`` is defined here but won't be equivalent if/until #549 is merged. As-is, ``dict`` would be slower due to boxing of the keys.

latkin replied:
The performance difference would be inconsequential, but it's up to you.

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +109,162 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+            | None ->
+                let opLength = op.Length
+                let sb = new System.Text.StringBuilder (opNamePrefix, opNamePrefix.Length + (opLength * maxOperatorNameLength))
+                for i = 0 to opLength - 1 do
```
With recent optimization PR merged last week, `for` loops over strings are now fast (not `IEnumerable`).  So for clarity I would recommend just using `for c in op do ... `

jack-pappas replied:
Done.

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +109,160 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+
+    // +++ GLOBAL STATE
+    /// Compiles a custom operator into a mangled operator name.
+    /// For example, "!%" becomes "op_DereferencePercent"
```
This is the exact same comment as on `compileCustomOpName` - should make it clear how the behaviors/assumptions differ

jack-pappas replied:
Done.

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +109,160 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+            match found with
+            | Some res -> res
+            | None ->
+                let rec loop (remaining : string) =
```
If you are looking to micro-optimize, this loop could return unit, and pass along a StringBuilder which is built up with the decompiled name.  That would make it tail-recursive (compiler will optimize into imperative loop), avoid some of the `Some` allocations, and avoid string concats.

latkin commented:
Overall LGTM, just some minor suggestions on improving clarity

latkin commented:
@dsyme do you have a style preference for this repo regarding .NET method calls: with/without parens?  e.g. `resizeArray.Add x` vs `resizeArray.Add(x)`.  This PR favors the former.

dsyme commented:
@latkin There is no strict preference, but more recent code tends to use the former, so it is ok to move code in that direction.

jack-pappas commented:
@latkin I cleaned up the loop in the ``decompileCustomOpName`` function per your suggestion.

One other thing I noticed was that there seemed to be more GC's reported in the ``--times`` output when I switched that one bit of code to use the ``for .. in .. do`` syntax to iterate over the string (as compared to indexing into the string). Not sure why that is, so I've changed the code back for now.

I've squashed the commits so this PR can be merged whenever you're ready.

latkin commented:
@jack-pappas Great! Would be interesting to find out what's up with the string loop, but no worries in the meantime. I'll try to get this merged soon.

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -109,66 +110,185 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+                    System.Text.StringBuilder (maxPossibleOpCharCount)
+
+                // Start decompiling just after the operator prefix.
+                match decompile sb opNamePrefixLen with
```
My thought was to have this simply return `string` directly - instead of returning `Some(sb.ToString())` just return `sb.ToString()`, instead of `None` return `opName`

jack-pappas replied:
@latkin Good point. I've made the change and updated the PR.

latkin commented on src/fsharp/PrettyNaming.fs:
```diff
@@ -318,38 +482,29 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
-
+        // extract the logical name from any mangled name produced by MakeMemberDataAndMangledNameForMemberVal
+        if s.Length <= 4 then None else
+        let s = chopStringTo s '.'
```
Apologies, my earlier suggestion to use this simplification was bogus.  This now fails/causes crash for identifiers with `.` in them (e.g. backtick identifiers). I'll fix before merging...

latkin commented:
Looks like you are actively updating this. Let me know when it's done so I can sync locally and re-run tests.

jack-pappas commented:
@latkin I'm finished making changes if you're going to fix that one piece of code before merging. Or, I can fix it and update the PR -- just let me know.

latkin commented:
Got it, re-running now.  I'll take care of the TryChopPropertyName thing, sorry about that.

latkin commented:
Applied to OOB branch
